### PR TITLE
[FEATURE] OpenPepXL: more efficient memory usage

### PR DESCRIPTION
**Pull request by enetz**
- Improved memory efficiency by using pointers: Enumeration of candidates of peptide pairs now points to the list of single peptides instead of copying the AASequences for each pair.

- Reduced memory usage by reworking the parallelization: previously the parallelized loop was looping over the spectra. The lists of candidates and results for each spectrum had to be enumerated and computed for each spectrum in parallel and this increased the memory usage proportional to the number of used threads. Now several parallelized loops loop over the candidate enumeration and result computation for one spectrum. This makes the parallelization slightly less efficient, but it removes the increase in memory usage for multiple threads.

I tested it on a data set with a 256 protein database, where some spectra have more than 10 million peptide pair candidates within a 6 ppm precursor tolerance window. I previously needed somewhere between 75GB and 100GB memory to run it on 50 threads on a cluster (2GB for each thread does not sound much, until you have dozens of them :) ). Now this data set needs around 5GB.

### CODE REVIEW

File: src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp [hunk 1]
```diff
@@ -549,8 +562,9 @@ namespace OpenMS
             cross_link_candidate.cross_linker_mass = cross_link_mass;
             cross_link_candidate.cross_linker_name = cross_link_name;
             cross_link_candidate.precursor_correction = precursor_corrections[i];
-            cross_link_candidates.push_back(cross_link_candidate);
 
+#pragma omp critical (cross_link_candidates_access)
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp]: again, all those criticals can be avoided if using a private vector per thread

File: src/openms/source/ANALYSIS/XLMS/OpenPepXLAlgorithm.cpp [hunk 2]
```diff
@@ -429,9 +429,9 @@ using namespace OpenMS;
     progresslogger.startProgress(0, 1, "Matching to theoretical spectra and scoring...");
     Size spectrum_counter = 0;
 
-#ifdef _OPENMP
-#pragma omp parallel for schedule(guided)
-#endif
+// #ifdef _OPENMP
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/XLMS/OpenPepXLAlgorithm.cpp]: the omp pragmas below can go as well since this is a serial region now, right?

**Comment by cbielow** [comment]
another hint: [guided](https://stackoverflow.com/questions/42970700/openmp-dynamic-vs-guided-scheduling) might not always be a good idea. You might be better off using dynamic.

**Comment by timosachsenberg** [comment]
@cbielow thanks for the valuable insights. I'll merge this PR as it passes CI.

**Comment by enetz** [comment]
I will try out the suggestions and open another PR later.

**Pull request merged**